internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go

```diff
@@ -0,0 +1,148 @@
+package builder
+
+import (
+	"github.com/hashicorp/consul/agent/connect"
+	"github.com/hashicorp/consul/internal/mesh/internal/types/intermediate"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1/pbproxystate"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+func (b *Builder) BuildDestinations(destinations []*intermediate.Destination) *Builder {
+	for _, destination := range destinations {
+		if destination.Explicit != nil {
+			b.buildExplicitDestination(destination)
+		}
+	}
+
+	return b
+}
+
+func (b *Builder) buildExplicitDestination(destination *intermediate.Destination) *Builder {
+	clusterName := DestinationClusterName(destination.Explicit.DestinationRef, destination.Explicit.Datacenter, b.trustDomain)
+	statPrefix := DestinationStatPrefix(destination.Explicit.DestinationRef, destination.Explicit.Datacenter)
+
+	// All endpoints should have the same protocol as the endpoints controller ensures that is the case,
+	// so it's sufficient to read just the first endpoint.
+	if len(destination.ServiceEndpoints.Endpoints.Endpoints) > 0 {
+		// Get destination port so that we can configure this destination correctly based on its protocol.
+		destPort := destination.ServiceEndpoints.Endpoints.Endpoints[0].Ports[destination.Explicit.DestinationPort]
+
+		// Find the destination proxy's port.
+		// Endpoints refs will need to route to mesh port instead of the destination port as that
+		// is the port of the destination's proxy.
+		meshPortName := findMeshPort(destination.ServiceEndpoints.Endpoints.Endpoints[0].Ports)
+
+		if destPort != nil {
+			return b.addOutboundDestinationListener(destination.Explicit).
+				addRouter(clusterName, statPrefix, destPort.Protocol).
```

Comment:
This feels a bit brittle given that it relies on the call order (listener before the router). I'm assuming the overlap is going to grow once RB's work is merged, and we don't know the cluster name yet.

We might want to consider merging these or considering a slightly different setup at some point.

Reply:
Yeah that's true. I'll merge the router bit to make it clear that it's coupled to listener.

Reply:
You could have the builder functions that create/stop a listener actually return a `ListenerBuilder` that internally tracks the index of the listener and has a pointer to the parent Builder.